File: Core/AppMode.swift
/// A named controller profile that maps buttons to actions.
///
/// Conforming types define the complete behaviour of the controller while a
/// particular application is in the foreground. Adding support for a new app
/// means creating a new struct that conforms to this protocol — nothing else
/// in the codebase needs to change.
///
/// ## Conformance
///
/// At minimum, provide `id`, `displayName`, `sfSymbolName`, and `buttonMap`.
/// The `comboMap` has a default empty implementation so it can be omitted
/// until you need it.
///
/// ```swift
/// struct WordMode: AppMode {
///     let id = "word"
///     let displayName = "Word"
///     let sfSymbolName = "doc.text"
///     var buttonMap: [GamepadButton: Action] { ... }
/// }
/// ```
protocol AppMode {

    /// A stable, URL-safe identifier. Used for serialising preferences to disk.
    var id: String { get }

    /// The name shown in the overlay toast and the menu bar.
    var displayName: String { get }

    /// An SF Symbol name for the mode icon shown in the overlay and menu bar.
    var sfSymbolName: String { get }

    /// Maps individual button presses to actions.
    ///
    /// Buttons not present in this dictionary produce no action when pressed.
    var buttonMap: [GamepadButton: Action] { get }

    /// Maps multi-button combos to actions.
    ///
    /// Combos take priority over single-button actions: if the current held-
    /// button set satisfies a combo here, only that combo's action fires.
    var comboMap: [ButtonCombo: Action] { get }
}

// MARK: - Default Implementations

extension AppMode {

    /// Modes do not define combos by default.
    var comboMap: [ButtonCombo: Action] { [:] }
}

File: buttonbridge/core/app_mode.py
"""AppMode abstract base class."""

from __future__ import annotations

from abc import ABC, abstractmethod

from .action import Action
from .button_combo import ButtonCombo
from .gamepad_button import GamepadButton


class AppMode(ABC):
    """
    A named controller profile that maps buttons to actions.

    Subclass this to add support for a new application.  Only
    ``id``, ``display_name``, ``sf_symbol_name``, and ``button_map``
    are required — ``combo_map`` defaults to empty.

    Example::

        class WordMode(AppMode):
            id           = "word"
            display_name = "Word"
            sf_symbol_name = "doc.text"   # used by the Swift overlay if ported back

            @property
            def button_map(self):
                return {
                    GamepadButton.A: keystroke("s", ["cmd"], id="word-save", name="Save"),
                    ...
                }
    """

    # Stable kebab-case identifier — used for persisting preferences to disk.
    @property
    @abstractmethod
    def id(self) -> str: ...

    # Name shown in the status bar and logs.
    @property
    @abstractmethod
    def display_name(self) -> str: ...

    # SF Symbol name (kept for future parity with any Swift UI layer).
    @property
    @abstractmethod
    def sf_symbol_name(self) -> str: ...

    # Maps individual button presses to actions.
    @property
    @abstractmethod
    def button_map(self) -> dict[GamepadButton, Action]: ...

    # Maps multi-button combos to actions.  Empty by default.
    @property
    def combo_map(self) -> dict[ButtonCombo, Action]:
        return {}

    def __repr__(self) -> str:
        return f"<{self.__class__.__name__} id={self.id!r}>"
